build/files.c

```diff
@@ -877,13 +886,29 @@ static VFA_t const virtualAttrs[] = {
  */
 static rpmRC parseForSimple(char * buf, FileEntry cur, ARGV_t * fileNames)
 {
-    char *s, *t;
+    char *s, *t, *end;
+    char *delim = " \t\n";
+    int quotes = 0;
     rpmRC res = RPMRC_OK;
     int allow_relative = (RPMFILE_PUBKEY|RPMFILE_DOC|RPMFILE_LICENSE);
 
     t = buf;
-    while ((s = strtokWithQuotes(t, " \t\n")) != NULL) {
```

Comment:
Not introduced or changed in this patch, but on a related note: it's more than a little hysterical that we split the line with tabs and spaces, forcing all this quoting on them, BUT then we don't allow more than one file per line anyway. Except for special %doc and %license.

So unless I'm missing something here, we could split by newline alone when allow_relative is false, and have spaces and tabs in filenames "just work" with zero quoting.

(ininitially commented this on a wrong PR)

Reply:
That's a good point and I don't think you're missing anything. In fact, splitting on newlines would simplify things quite a bit.

This PR currently assumes that backslash escaping of globs and spaces *is* useful, so it 1) makes it work properly and 2) uses it internally to implement proper quotation (simpler code & more flexible). However, the only real reason to support backslash escaping at all is to allow for a combination of spaces and globs, e.g. `/foo\ *`. A combination of non-globs and globs (`/foo\*[123]`) is much harder to implement as it depends on the hairy internals of `rpmGlob()` which are hard to fix without breaking everything else, so that's not in scope of this PR, anyway. That means, if we let go of backslashes completely *and* split on newlines, we could simplify the semantics as follows:

- No quotes mean globs will be interpreted, business as usual.
- Quotes mean do not interpret any globs.
- Spaces are always taken literally and don't need to be quoted.

This would conveniently take care of the space & glob use case, too, because you could then just write `/foo *`.

What do you think? I'm considering to give this a go in a separate topic branch and if it goes well, replace this one with it.

Reply:
Having slept over this... I stand corrected, we *do* support multiple files on a single line, just not for regular files (it's an [artificial limitation](https://github.com/rpm-software-management/rpm/blob/bb36aac905cf8eb470f085574641d92c8d8ca20f/build/files.c#L2563)). For special files (e.g. `%doc`), there may and often actually are[1] multiple files on a line, and they can and often do[1] use globs, therefore escaping needs to be possible either with quotes or backslashes, ideally both, to allow for space & glob combinations.

[1]
```sh
$ cd rpmspecs/  # from https://pkgs.fedoraproject.org/repo/rpm-specs-latest.tar.xz
$ grep '%doc' *.spec | wc -l
31613
```

[2]
```sh
$ grep '%doc' zsh.spec
%doc README LICENCE Etc/BUGS Etc/CONTRIBUTORS Etc/FAQ FEATURES MACHINES
%doc NEWS Etc/zsh-development-guide Etc/completion-style-guide
%doc Doc/*.html
```

Reply:
Hmm, I can see now that you mentioned "if allow_relative is false" which basically implies the above. Either way, once we support multiple files on a line, it doesn't make sense, IMO, to handle regular and non-regular files differently.

Reply:
Alternatively, we claim that
1) glob escaping only works for regular files and
2) regular files are separated by newlines and therefore don't need quoting for spaces
Which would make the patch a little simpler and less intrusive, otoh, it would make things inconsistent and more complicated to document.